narwhal/consensus/src/bullshark.rs

```diff
@@ -236,6 +237,21 @@ impl Bullshark {
         }
     }
 
+    // Returns the PublicKey of the authority which is the leader for the provided `round`.
+    // Pay attention that this method will return always the first authority as the leader
+    // when used under a test environment.
+    pub fn leader_authority(committee: &Committee, _round: Round) -> PublicKey {
+        cfg_if::cfg_if! {
+            if #[cfg(test)] {
```

Comment:
Ultimately I would like to completely remove this from here. This is very implicit and easily to miss in tests as well. Also wouldn't like to mix production with test code - I believe separate dependency injection should take place.

Reply:
Exactly, we never test the case where we change leaders. The certificates we produce in the tests are "perfect" certificates in the sense that each certificate always refers to all the parents of previous round. So in the end there might not be that much of a difference - but 100% agree that we should test with different leaders. I'll re-iterate in next PR